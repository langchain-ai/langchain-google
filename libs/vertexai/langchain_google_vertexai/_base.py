from __future__ import annotations

from concurrent.futures import Executor
from typing import Any, ClassVar, Dict, List, Optional

import vertexai  # type: ignore[import-untyped]
from google.api_core.client_options import ClientOptions
from google.cloud.aiplatform.gapic import (
    PredictionServiceAsyncClient,
    PredictionServiceClient,
)
from google.cloud.aiplatform.models import Prediction
from google.protobuf import json_format
from google.protobuf.struct_pb2 import Value
from langchain_core.outputs import Generation, LLMResult
from langchain_core.pydantic_v1 import BaseModel, Field, root_validator
from vertexai.language_models import (  # type: ignore[import-untyped]
    TextGenerationModel,
)
from vertexai.preview.language_models import (  # type: ignore[import-untyped]
    ChatModel as PreviewChatModel,
)
from vertexai.preview.language_models import (
    CodeChatModel as PreviewCodeChatModel,
)

from langchain_google_vertexai._enums import HarmBlockThreshold, HarmCategory
from langchain_google_vertexai._utils import (
    get_client_info,
    get_user_agent,
    is_codey_model,
    is_gemini_model,
)

_PALM_DEFAULT_MAX_OUTPUT_TOKENS = TextGenerationModel._DEFAULT_MAX_OUTPUT_TOKENS
_PALM_DEFAULT_TEMPERATURE = 0.0
_PALM_DEFAULT_TOP_P = 0.95
_PALM_DEFAULT_TOP_K = 40


class _VertexAIBase(BaseModel):
    client: Any = None  #: :meta private:
    project: Optional[str] = None
    "The default GCP project to use when making Vertex API calls."
    location: str = "us-central1"
    "The default location to use when making API calls."
    request_parallelism: int = 5
    "The amount of parallelism allowed for requests issued to VertexAI models. "
    "Default is 5."
    max_retries: int = 6
    """The maximum number of retries to make when generating."""
    task_executor: ClassVar[Optional[Executor]] = Field(default=None, exclude=True)
    stop: Optional[List[str]] = None
    "Optional list of stop words to use when generating."
    model_name: Optional[str] = None
    "Underlying model name."


class _VertexAICommon(_VertexAIBase):
    client_preview: Any = None  #: :meta private:
    model_name: str
    "Underlying model name."
    temperature: Optional[float] = None
    "Sampling temperature, it controls the degree of randomness in token selection."
    max_output_tokens: Optional[int] = None
    "Token limit determines the maximum amount of text output from one prompt."
    top_p: Optional[float] = None
    "Tokens are selected from most probable to least until the sum of their "
    "probabilities equals the top-p value. Top-p is ignored for Codey models."
    top_k: Optional[int] = None
    "How the model selects tokens for output, the next token is selected from "
    "among the top-k most probable tokens. Top-k is ignored for Codey models."
    credentials: Any = Field(default=None, exclude=True)
    "The default custom credentials (google.auth.credentials.Credentials) to use "
    "when making API calls. If not provided, credentials will be ascertained from "
    "the environment."
    n: int = 1
    """How many completions to generate for each prompt."""
    streaming: bool = False
    """Whether to stream the results or not."""
    safety_settings: Optional[Dict[HarmCategory, HarmBlockThreshold]] = None
    """The default safety settings to use for all generations. 
    
        For example: 

            from langchain_google_vertexai import HarmBlockThreshold, HarmCategory

            safety_settings = {
                HarmCategory.HARM_CATEGORY_UNSPECIFIED: HarmBlockThreshold.BLOCK_NONE,
                HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_ONLY_HIGH,
                HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_LOW_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,
            }
            """  # noqa: E501

    @property
    def _llm_type(self) -> str:
        return "vertexai"

    @property
    def is_codey_model(self) -> bool:
        return is_codey_model(self.model_name)

    @property
    def _is_gemini_model(self) -> bool:
        return is_gemini_model(self.model_name)

    @property
    def _identifying_params(self) -> Dict[str, Any]:
        """Gets the identifying parameters."""
        return {**{"model_name": self.model_name}, **self._default_params}

    @property
    def _default_params(self) -> Dict[str, Any]:
        if self._is_gemini_model:
            default_params = {}
        else:
            default_params = {
                "temperature": _PALM_DEFAULT_TEMPERATURE,
                "max_output_tokens": _PALM_DEFAULT_MAX_OUTPUT_TOKENS,
                "top_p": _PALM_DEFAULT_TOP_P,
                "top_k": _PALM_DEFAULT_TOP_K,
            }
        params = {
            "temperature": self.temperature,
            "max_output_tokens": self.max_output_tokens,
            "candidate_count": self.n,
        }
        if not self.is_codey_model:
            params.update(
                {
                    "top_k": self.top_k,
                    "top_p": self.top_p,
                }
            )
        updated_params = {}
        for param_name, param_value in params.items():
            default_value = default_params.get(param_name)
            if param_value or default_value:
                updated_params[param_name] = (
                    param_value if param_value else default_value
                )
        return updated_params

    @property
    def _user_agent(self) -> str:
        """Gets the User Agent."""
        _, user_agent = get_user_agent(f"{type(self).__name__}_{self.model_name}")
        return user_agent

    @classmethod
    def _init_vertexai(cls, values: Dict) -> None:
        vertexai.init(
            project=values.get("project"),
            location=values.get("location"),
            credentials=values.get("credentials"),
        )
        return None

    def _prepare_params(
        self,
        stop: Optional[List[str]] = None,
        stream: bool = False,
        **kwargs: Any,
    ) -> dict:
        stop_sequences = stop or self.stop
        params_mapping = {"n": "candidate_count"}
        params = {params_mapping.get(k, k): v for k, v in kwargs.items()}
        params = {**self._default_params, "stop_sequences": stop_sequences, **params}
        if stream or self.streaming:
            params.pop("candidate_count")
        return params

    def get_num_tokens(self, text: str) -> int:
        """Get the number of tokens present in the text.

        Useful for checking if an input will fit in a model's context window.

        Args:
            text: The string input to tokenize.

        Returns:
            The integer number of tokens in the text.
        """
        is_palm_chat_model = isinstance(
            self.client_preview, PreviewChatModel
        ) or isinstance(self.client_preview, PreviewCodeChatModel)
        if is_palm_chat_model:
            result = self.client_preview.start_chat().count_tokens(text)
        else:
            result = self.client_preview.count_tokens([text])

        return result.total_tokens


class _BaseVertexAIModelGarden(_VertexAIBase):
    """Large language models served from Vertex AI Model Garden."""

    async_client: Any = None  #: :meta private:
    endpoint_id: str
    "A name of an endpoint where the model has been deployed."
    allowed_model_args: Optional[List[str]] = None
    "Allowed optional args to be passed to the model."
    prompt_arg: str = "prompt"
    result_arg: Optional[str] = "generated_text"
    "Set result_arg to None if output of the model is expected to be a string."
    "Otherwise, if it's a dict, provided an argument that contains the result."
    single_example_per_request: bool = True
    "LLM endpoint currently serves only the first example in the request"

    @root_validator()
    def validate_environment(cls, values: Dict) -> Dict:
        """Validate that the python package exists in environment."""

        if not values["project"]:
            raise ValueError(
                "A GCP project should be provided to run inference on Model Garden!"
            )

        client_options = ClientOptions(
            api_endpoint=f"{values['location']}-aiplatform.googleapis.com"
        )
        client_info = get_client_info(module="vertex-ai-model-garden")
        values["client"] = PredictionServiceClient(
            client_options=client_options, client_info=client_info
        )
        values["async_client"] = PredictionServiceAsyncClient(
            client_options=client_options, client_info=client_info
        )
        return values

    @property
    def endpoint_path(self) -> str:
        return self.client.endpoint_path(
            project=self.project, location=self.location, endpoint=self.endpoint_id
        )

    @property
    def _llm_type(self) -> str:
        return "vertexai_model_garden"

    def _prepare_request(self, prompts: List[str], **kwargs: Any) -> List["Value"]:
        instances = []
        for prompt in prompts:
            if self.allowed_model_args:
                instance = {
                    k: v for k, v in kwargs.items() if k in self.allowed_model_args
                }
            else:
                instance = {}
            instance[self.prompt_arg] = prompt
            instances.append(instance)

        predict_instances = [
            json_format.ParseDict(instance_dict, Value()) for instance_dict in instances
        ]
        return predict_instances

    def _parse_response(self, predictions: "Prediction") -> LLMResult:
        generations: List[List[Generation]] = []
        for result in predictions.predictions:
            if isinstance(result, str):
                generations.append([Generation(text=self._parse_prediction(result))])
            else:
                generations.append(
                    [
                        Generation(text=self._parse_prediction(prediction))
                        for prediction in result
                    ]
                )
        return LLMResult(generations=generations)

    def _parse_prediction(self, prediction: Any) -> str:
        if isinstance(prediction, str):
            return prediction

        if self.result_arg:
            try:
                return prediction[self.result_arg]
            except KeyError:
                if isinstance(prediction, str):
                    error_desc = (
                        "Provided non-None `result_arg` (result_arg="
                        f"{self.result_arg}). But got prediction of type "
                        f"{type(prediction)} instead of dict. Most probably, you"
                        "need to set `result_arg=None` during VertexAIModelGarden "
                        "initialization."
                    )
                    raise ValueError(error_desc)
                else:
                    raise ValueError(f"{self.result_arg} key not found in prediction!")

        return prediction
