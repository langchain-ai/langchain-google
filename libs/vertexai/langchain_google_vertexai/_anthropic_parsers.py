from typing import Any, List, Optional, Type

from langchain_core.messages import AIMessage, ToolCall
from langchain_core.messages.tool import tool_call
from langchain_core.output_parsers import BaseGenerationOutputParser
from langchain_core.outputs import ChatGeneration, Generation
from pydantic import BaseModel, ConfigDict


class ToolsOutputParser(BaseGenerationOutputParser):
    first_tool_only: bool = False
    args_only: bool = False
    pydantic_schemas: Optional[List[Type[BaseModel]]] = None

    model_config = ConfigDict(
        extra="forbid",
    )

    def parse_result(self, result: List[Generation], *, partial: bool = False) -> Any:
        """Parse a list of candidate model Generations into a specific format.

        Args:
            result: A list of Generations to be parsed. The Generations are assumed
                to be different candidate outputs for a single model input.

        Returns:
            Structured output.
        """
        if not result or not isinstance(result[0], ChatGeneration):
            return None if self.first_tool_only else []

        message = result[0].message
        tool_calls: List[Any] = []

        if isinstance(message, AIMessage) and message.tool_calls:
            tool_calls = message.tool_calls
        elif isinstance(message.content, list):
            content: Any = message.content
            tool_calls = _extract_tool_calls(content)

        if self.pydantic_schemas:
            tool_calls = [self._pydantic_parse(tc) for tc in tool_calls]
        elif self.args_only:
            tool_calls = [tc["args"] for tc in tool_calls]

        if self.first_tool_only:
            return tool_calls[0] if tool_calls else None
        else:
            return [tool_call for tool_call in tool_calls]

    def _pydantic_parse(self, tool_call: dict) -> BaseModel:
        cls_ = {schema.__name__: schema for schema in self.pydantic_schemas or []}[
            tool_call["name"]
        ]
        return cls_(**tool_call["args"])


def _extract_tool_calls(content: List[dict]) -> List[ToolCall]:
    tool_calls = []
    for block in content:
        if block["type"] == "tool_use":
            tool_calls.append(
                tool_call(name=block["name"], args=block["input"], id=block["id"])
            )
    return tool_calls
