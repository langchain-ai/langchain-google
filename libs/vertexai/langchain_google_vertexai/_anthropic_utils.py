import re
from typing import Dict, List, Optional, Tuple, Union

from langchain_core.messages import BaseMessage

_message_type_lookups = {"human": "user", "ai": "assistant"}


def _format_image(image_url: str) -> Dict:
    """Formats a message image to a dict for anthropic api."""
    regex = r"^data:(?P<media_type>image/.+);base64,(?P<data>.+)$"
    match = re.match(regex, image_url)
    if match is None:
        raise ValueError(
            "Anthropic only supports base64-encoded images currently."
            " Example: data:image/png;base64,'/9j/4AAQSk'..."
        )
    return {
        "type": "base64",
        "media_type": match.group("media_type"),
        "data": match.group("data"),
    }


def _format_messages_anthropic(
    messages: List[BaseMessage],
) -> Tuple[Optional[str], List[Dict]]:
    """Formats messages for anthropic."""
    system_message: Optional[str] = None
    formatted_messages: List[Dict] = []

    for i, message in enumerate(messages):
        if message.type == "system":
            if i != 0:
                raise ValueError("System message must be at beginning of message list.")
            if not isinstance(message.content, str):
                raise ValueError(
                    "System message must be a string, "
                    f"instead was: {type(message.content)}"
                )
            system_message = message.content
            continue

        role = _message_type_lookups[message.type]
        content: Union[str, List[Dict]]

        if not isinstance(message.content, str):
            # parse as dict
            assert isinstance(
                message.content, list
            ), "Anthropic message content must be str or list of dicts"

            # populate content
            content = []
            for item in message.content:
                if isinstance(item, str):
                    content.append(
                        {
                            "type": "text",
                            "text": item,
                        }
                    )
                elif isinstance(item, dict):
                    if "type" not in item:
                        raise ValueError("Dict content item must have a type key")
                    elif item["type"] == "image_url":
                        # convert format
                        source = _format_image(item["image_url"]["url"])
                        content.append(
                            {
                                "type": "image",
                                "source": source,
                            }
                        )
                    elif item["type"] == "tool_use":
                        item.pop("text", None)
                        content.append(item)
                    else:
                        content.append(item)
                else:
                    raise ValueError(
                        f"Content items must be str or dict, instead was: {type(item)}"
                    )
        else:
            content = message.content

        formatted_messages.append(
            {
                "role": role,
                "content": content,
            }
        )
    return system_message, formatted_messages
