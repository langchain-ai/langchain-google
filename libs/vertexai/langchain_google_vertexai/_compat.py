"""Go from v1 content blocks to VertexAI format."""

from typing import Any, cast

from langchain_core.messages import content as types


def _convert_from_v1_to_vertex(
    content: list[types.ContentBlock], model_provider: str | None
) -> list[dict[str, Any]]:
    """Convert v1 content blocks to VertexAI content.

    Args:
        content: List of v1 `ContentBlock` objects.
        model_provider: The model provider name that generated the v1 content.

    Returns:
        List of dictionaries in VertexAI content format.
    """
    new_content: list = []
    for block in content:
        block_dict = dict(block)  # (For typing)

        # TextContentBlock
        if block_dict["type"] == "text":
            new_block = {"type": "text", "text": block_dict.get("text", "")}
            new_content.append(new_block)
            # Citations are only handled on output. Can't pass them back :/

        # ReasoningContentBlock -> thinking
        elif block_dict["type"] == "reasoning" and model_provider == "google_vertexai":
            # Google requires passing back the thought_signature when available.
            # Signatures are only provided when function calling is enabled.
            if "extras" in block_dict and isinstance(block_dict["extras"], dict):
                extras = block_dict["extras"]
                if "signature" in extras:
                    new_block = {
                        "thought": True,
                        "text": block_dict.get("reasoning", ""),
                        "thought_signature": extras["signature"],
                    }
                    new_content.append(new_block)
                # else: skip reasoning blocks without signatures
                # TODO: log a warning?
            # else: skip reasoning blocks without extras
            # TODO: log a warning?

        # ToolCall -> FunctionCall
        elif block_dict["type"] == "tool_call":
            # read from .tool_calls
            continue

        elif block_dict["type"] == "server_tool_call":
            if block_dict.get("name") == "code_interpreter":
                # LangChain v0 format
                args = cast(dict, block_dict.get("args", {}))
                executable_code = {
                    "type": "executable_code",
                    "executable_code": args.get("code", ""),
                    "language": args.get("language", ""),
                    "id": block_dict.get("id", ""),
                }
                # Google generativelanguage format
                new_content.append(executable_code)

        elif block_dict["type"] == "server_tool_result":
            extras = cast(dict, block_dict.get("extras", {}))
            if extras.get("block_type") == "code_execution_result":
                # LangChain v0 format
                code_execution_result = {
                    "type": "code_execution_result",
                    "code_execution_result": block_dict.get("output", ""),
                    "outcome": extras.get("outcome", ""),
                    "tool_call_id": block_dict.get("tool_call_id", ""),
                }
                # Google generativelanguage format
                new_content.append(code_execution_result)

        elif block_dict["type"] == "non_standard":
            new_content.append(block_dict["value"])

    return new_content
