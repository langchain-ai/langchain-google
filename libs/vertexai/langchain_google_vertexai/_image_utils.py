from __future__ import annotations

import base64
import os
import re
from typing import Dict, Union
from urllib.parse import urlparse

import requests
from google.cloud import storage


class ImageBytesLoader:
    """Loads image bytes from multiple sources given a string.

    Currently supported:
        - Google cloud storage URI
        - B64 Encoded image string
        - Local file path
        - URL
    """

    def __init__(
        self,
        project: Union[str, None] = None,
    ) -> None:
        """Constructor

        Args:
            project: Google Cloud project id. Defaults to none.
        """
        self._project = project

    def load_bytes(self, image_string: str) -> bytes:
        """Routes to the correct loader based on the image_string.

        Args:
            image_string: Can be either:
                    - Google cloud storage URI
                    - B64 Encoded image string
                    - Local file path
                    - URL

        Returns:
            Image bytes.
        """

        if image_string.startswith("gs://"):
            return self._bytes_from_gcs(image_string)

        if image_string.startswith("data:image/"):
            return self._bytes_from_b64(image_string)

        if self._is_url(image_string):
            return self._bytes_from_url(image_string)

        if os.path.exists(image_string):
            return self._bytes_from_file(image_string)

        raise ValueError(
            "Image string must be one of: Google Cloud Storage URI, "
            "b64 encoded image string (data:image/...), valid image url, "
            f"or existing local image file. Instead got '{image_string}'."
        )

    def _bytes_from_b64(self, base64_image: str) -> bytes:
        """Gets image bytes from a base64 encoded string.

        Args:
            base64_image: Encoded image in b64 format.

        Returns:
            Image bytes
        """

        pattern = r"data:image/\w{2,4};base64,(.*)"
        match = re.search(pattern, base64_image)

        if match is not None:
            encoded_string = match.group(1)
            return base64.b64decode(encoded_string)

        raise ValueError(f"Error in b64 encoded image. Must follow pattern: {pattern}")

    def _bytes_from_file(self, file_path: str) -> bytes:
        """Gets image bytes from a local file path.

        Args:
            file_path: Existing file path.

        Returns:
            Image bytes
        """
        with open(file_path, "rb") as image_file:
            image_bytes = image_file.read()
        return image_bytes

    def _bytes_from_url(self, url: str) -> bytes:
        """Gets image bytes from a public url.

        Args:
            url: Valid url.

        Raises:
            HTTP Error if there is one.

        Returns:
            Image bytes
        """

        response = requests.get(url)

        if not response.ok:
            response.raise_for_status()

        return response.content

    def _bytes_from_gcs(self, gcs_uri: str) -> bytes:
        """Gets image bytes from a Google Cloud Storage uri.

        Args:
            gcs_uri: Valid gcs uri.

        Raises:
            ValueError if there are more than one blob matching the uri.

        Returns:
            Image bytes
        """

        gcs_client = storage.Client(project=self._project)
        return storage.Blob.from_string(gcs_uri, gcs_client).download_as_bytes()

    def _is_url(self, url_string: str) -> bool:
        """Checks if a url is valid.

        Args:
            url_string: Url to check.

        Returns:
            Whether the url is valid.
        """
        try:
            result = urlparse(url_string)
            return all([result.scheme, result.netloc])
        except Exception:
            return False


def image_bytes_to_b64_string(
    image_bytes: bytes, encoding: str = "ascii", image_format: str = "png"
) -> str:
    """Encodes image bytes into a b64 encoded string.

    Args:
        image_bytes: Bytes of the image.
        encoding: Type of encoding in the string. 'ascii' by default.
        image_format: Format of the image. 'png' by default.

    Returns:
        B64 image encoded string.
    """
    encoded_bytes = base64.b64encode(image_bytes).decode(encoding)
    return f"data:image/{image_format};base64,{encoded_bytes}"


def create_text_content_part(message_str: str) -> Dict:
    """Create a dictionary that can be part of a message content list.

    Args:
        message_str: Message as an string.

    Returns:
        Dictionary that can be part of a message content list.
    """
    return {"type": "text", "text": message_str}


def create_image_content_part(image_str: str) -> Dict:
    """Create a dictionary that can be part of a message content list.

    Args:
        image_str: Can be either:
            - b64 encoded image data
            - GCS uri
            - Url
            - Path to an image.

    Returns:
        Dictionary that can be part of a message content list.
    """
    return {"type": "image_url", "image_url": {"url": image_str}}


def get_image_str_from_content_part(content_part: str | Dict) -> str | None:
    """Parses an image string from a dictionary with the correct format.

    Args:
        content_part: String or dictionary.

    Returns:
        Image string if the dictionary has the correct format otherwise None.
    """

    if isinstance(content_part, str):
        return None

    if content_part.get("type") != "image_url":
        return None

    image_str = content_part.get("image_url", {}).get("url")

    if isinstance(image_str, str):
        return image_str
    else:
        return None


def get_text_str_from_content_part(content_part: str | Dict) -> str | None:
    """Parses an string from a dictionary or string with the correct format.

    Args:
        content_part:  String or dictionary.

    Returns:
        String if the dictionary has the correct format or the input is an string,
        otherwise None.
    """

    if isinstance(content_part, str):
        return content_part

    if content_part.get("type") != "text":
        return None

    text = content_part.get("text")

    if isinstance(text, str):
        return text
    else:
        return None
