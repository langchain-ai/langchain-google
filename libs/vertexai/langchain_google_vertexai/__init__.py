from google.cloud.aiplatform_v1beta1.types import (
    FunctionCallingConfig,
    FunctionDeclaration,
    Schema,
    ToolConfig,
    Type,
)

from langchain_google_vertexai._enums import HarmBlockThreshold, HarmCategory
from langchain_google_vertexai.chains import create_structured_runnable
from langchain_google_vertexai.chat_models import ChatVertexAI
from langchain_google_vertexai.embeddings import VertexAIEmbeddings
from langchain_google_vertexai.functions_utils import (
    PydanticFunctionsOutputParser,
)
from langchain_google_vertexai.gemma import (
    GemmaChatLocalHF,
    GemmaChatLocalKaggle,
    GemmaChatVertexAIModelGarden,
    GemmaLocalHF,
    GemmaLocalKaggle,
    GemmaVertexAIModelGarden,
)
from langchain_google_vertexai.llms import VertexAI
from langchain_google_vertexai.model_garden import VertexAIModelGarden
from langchain_google_vertexai.vectorstores import (
    DataStoreDocumentStorage,
    GCSDocumentStorage,
    VectorSearchVectorStore,
    VectorSearchVectorStoreDatastore,
    VectorSearchVectorStoreGCS,
)
from langchain_google_vertexai.vision_models import (
    VertexAIImageCaptioning,
    VertexAIImageCaptioningChat,
    VertexAIImageEditorChat,
    VertexAIImageGeneratorChat,
    VertexAIVisualQnAChat,
)

__all__ = [
    "ChatVertexAI",
    "create_structured_runnable",
    "DataStoreDocumentStorage",
    "FunctionCallingConfig",
    "FunctionDeclaration",
    "GCSDocumentStorage",
    "GemmaChatLocalHF",
    "GemmaChatLocalKaggle",
    "GemmaChatVertexAIModelGarden",
    "GemmaLocalHF",
    "GemmaLocalKaggle",
    "GemmaVertexAIModelGarden",
    "HarmBlockThreshold",
    "HarmCategory",
    "PydanticFunctionsOutputParser",
    "Schema",
    "ToolConfig",
    "Type",
    "VectorSearchVectorStore",
    "VectorSearchVectorStoreDatastore",
    "VectorSearchVectorStoreGCS",
    "VertexAI",
    "VertexAIEmbeddings",
    "VertexAIImageCaptioning",
    "VertexAIImageCaptioningChat",
    "VertexAIImageEditorChat",
    "VertexAIImageGeneratorChat",
    "VertexAIModelGarden",
    "VertexAIVisualQnAChat",
]
