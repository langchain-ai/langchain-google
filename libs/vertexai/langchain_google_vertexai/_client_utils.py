import asyncio
from functools import lru_cache
from typing import Any, Callable, Literal, Optional, Tuple, Union
from weakref import WeakKeyDictionary

from google.api_core.client_options import ClientOptions
from google.cloud.aiplatform_v1.services.prediction_service import (
    PredictionServiceAsyncClient as v1PredictionServiceAsyncClient,
)
from google.cloud.aiplatform_v1.services.prediction_service import (
    PredictionServiceClient as v1PredictionServiceClient,
)
from google.cloud.aiplatform_v1beta1.services.prediction_service import (
    PredictionServiceAsyncClient as v1beta1PredictionServiceAsyncClient,
)
from google.cloud.aiplatform_v1beta1.services.prediction_service import (
    PredictionServiceClient as v1beta1PredictionServiceClient,
)

from langchain_google_vertexai._utils import get_client_info


@lru_cache
def _get_client_options(
    api_endpoint: Optional[str],
    cert_source: Optional[Callable[[], Tuple[bytes, bytes]]],
) -> ClientOptions:
    """Return a shared ClientOptions object for each unique configuration."""
    client_options = ClientOptions(api_endpoint=api_endpoint)
    if cert_source:
        client_options.client_cert_source = cert_source
    return client_options


# Cache sync client
@lru_cache
def _get_prediction_client(
    *,
    endpoint_version: Literal["v1", "v1beta1"],
    credentials: Any,
    client_options: ClientOptions,
    transport: str,
    user_agent: str,
) -> Union[v1PredictionServiceClient, v1beta1PredictionServiceClient]:
    """Return a shared PredictionServiceClient."""
    client_kwargs: dict[str, Any] = {
        "credentials": credentials,
        "client_options": client_options,
        "client_info": get_client_info(module=user_agent),
        "transport": transport,
    }
    if endpoint_version == "v1":
        return v1PredictionServiceClient(**client_kwargs)
    else:
        return v1beta1PredictionServiceClient(**client_kwargs)


# Cache async client - must store caches per event loop
_client_caches: WeakKeyDictionary = WeakKeyDictionary()


def _create_async_prediction_client(
    *,
    endpoint_version: Literal["v1", "v1beta1"],
    credentials: Any,
    client_options: ClientOptions,
    user_agent: str,
) -> Union[v1PredictionServiceAsyncClient, v1beta1PredictionServiceAsyncClient]:
    """Create a new PredictionServiceAsyncClient."""
    async_client_kwargs: dict[str, Any] = {
        "client_options": client_options,
        "client_info": get_client_info(module=user_agent),
        "credentials": credentials,
        # async clients don't support "rest" transport
        # https://github.com/googleapis/gapic-generator-python/issues/1962
        "transport": "grpc_asyncio",
    }

    if endpoint_version == "v1":
        return v1PredictionServiceAsyncClient(**async_client_kwargs)
    else:
        return v1beta1PredictionServiceAsyncClient(**async_client_kwargs)


def _get_async_prediction_client(
    *,
    endpoint_version: Literal["v1", "v1beta1"],
    credentials: Any,
    client_options: ClientOptions,
    user_agent: str,
):
    """Return a shared PredictionServiceAsyncClient per event loop."""
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        # If no event loop is running, don't cache
        return _create_async_prediction_client(
            endpoint_version=endpoint_version,
            credentials=credentials,
            client_options=client_options,
            user_agent=user_agent,
        )

    # Get or create cache for this event loop
    if loop not in _client_caches:
        _client_caches[loop] = {}

    cache_key = (endpoint_version, id(credentials), id(client_options), user_agent)

    if cache_key not in _client_caches[loop]:
        _client_caches[loop][cache_key] = _create_async_prediction_client(
            endpoint_version=endpoint_version,
            credentials=credentials,
            client_options=client_options,
            user_agent=user_agent,
        )

    return _client_caches[loop][cache_key]
