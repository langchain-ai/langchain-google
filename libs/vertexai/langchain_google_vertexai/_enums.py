from vertexai.generative_models import (  # type: ignore
    HarmBlockThreshold,
    HarmCategory,
    SafetySetting,
)

__all__ = ["HarmBlockThreshold", "HarmCategory", "SafetySetting"]
