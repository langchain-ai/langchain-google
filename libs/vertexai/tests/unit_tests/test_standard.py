from typing import Type

from langchain_core.language_models import BaseChatModel
from langchain_standard_tests.unit_tests import ChatModelUnitTests

from langchain_google_vertexai import ChatVertexAI


class TestGeminiAIStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[BaseChatModel]:
        return ChatVertexAI

    @property
    def chat_model_params(self) -> dict:
        return {"model_name": "gemini-1.0-pro-001"}


class TestGemini_15_AIStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[BaseChatModel]:
        return ChatVertexAI

    @property
    def chat_model_params(self) -> dict:
        return {"model_name": "gemini-1.5-pro-001"}
