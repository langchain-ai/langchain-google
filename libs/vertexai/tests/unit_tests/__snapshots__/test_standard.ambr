# serializer version: 1
# name: TestGeminiAIStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain',
      'chat_models',
      'vertexai',
      'ChatVertexAI',
    ]),
    'kwargs': dict({
      'default_metadata': list([
      ]),
      'full_model_name': 'projects/test-project/locations/us-central1/publishers/google/models/gemini-1.0-pro-001',
      'location': 'us-central1',
      'max_output_tokens': 100,
      'max_retries': 2,
      'model_family': '1',
      'model_name': 'gemini-1.0-pro-001',
      'n': 1,
      'project': 'test-project',
      'request_parallelism': 5,
      'stop': list([
      ]),
      'temperature': 0.0,
    }),
    'lc': 1,
    'name': 'ChatVertexAI',
    'type': 'constructor',
  })
# ---
# name: TestGemini_15_AIStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain',
      'chat_models',
      'vertexai',
      'ChatVertexAI',
    ]),
    'kwargs': dict({
      'default_metadata': list([
      ]),
      'full_model_name': 'projects/test-project/locations/us-central1/publishers/google/models/gemini-1.5-pro-001',
      'location': 'us-central1',
      'max_output_tokens': 100,
      'max_retries': 2,
      'model_family': '2',
      'model_name': 'gemini-1.5-pro-001',
      'n': 1,
      'project': 'test-project',
      'request_parallelism': 5,
      'stop': list([
      ]),
      'temperature': 0.0,
    }),
    'lc': 1,
    'name': 'ChatVertexAI',
    'type': 'constructor',
  })
# ---
